qiskit/transpiler/passes/optimization/inverse_cancellation.py

```diff
@@ -111,19 +111,20 @@ def _run_on_self_inverse(self, dag: DAGCircuit):
             for gate_cancel_run in gate_runs:
                 partitions = []
                 chunk = []
-                for i in range(len(gate_cancel_run) - 1):
+                for i in range(len(gate_cancel_run)):
                     if gate_cancel_run[i].op == gate:
                         chunk.append(gate_cancel_run[i])
                     else:
                         if chunk:
                             partitions.append(chunk)
                             chunk = []
                         continue
-                    if gate_cancel_run[i].qargs != gate_cancel_run[i + 1].qargs:
+                    if (
+                        i == len(gate_cancel_run) - 1
+                        or gate_cancel_run[i].qargs != gate_cancel_run[i + 1].qargs
+                    ):
                         partitions.append(chunk)
                         chunk = []
-                chunk.append(gate_cancel_run[-1])
```

Comment:
Personally I think I'd probably have just done:

```python
if gate_cancel_run[-1].op == gate:
    chunk.append(gate_cancel_run[-1])
```
for a minimal change instead of changing the loop and if condition above as that's more natural way to read this logic to me, but what you've got should work I think.

Reply:
I agree that by avoiding a complex condition, like how you did, it is way more readable, but note that this change is not enough and it should be 

```python
if gate_cancel_run[-1].op == gate:
    chunk.append(gate_cancel_run[-1])
# now we are sure anymore that whether the chuck is empty or not
if chunk:
  partitions.append(chunk)
```

All of these conditions together are somehow similar to 

https://github.com/Qiskit/qiskit/blob/5af02486b45347570c5e5c42a22ef64791a27add/qiskit/transpiler/passes/optimization/inverse_cancellation.py#L115-L119

so I decided to make the last item a part of the loop. Still if you think that is a better way, I can change it!

Reply:
I apologize apparently a comment I wrote I must have forgotten to hit the add to review button on (or it got lost somehow). This was my idle musing comment that I mention in the top level review comment, and not the one I intended as a realistic suggestion. I left the inline comment again on a more concrete suggestion.